Kuzzle/Offline/OfflineManager.cs

```diff
@@ -0,0 +1,200 @@
+    /// </summary>
+    public override int MaxRequestDelay {
+      get { return maxRequestDelay; }
+      set { maxRequestDelay = value < 0 ? 0 : value; }
+    }
+
+    public override Func<JObject, bool> QueueFilter {
+      get { return queueFilter; }
+      set { queueFilter = value ?? ((obj) => true); }
+    }
+
+    /// <summary>
+    /// Queue requests when network is down,
+    /// and automatically replay them when the SDK successfully reconnects.
+    /// </summary>
+    public override bool AutoRecover { get; set; } = true;
+
+    internal OfflineManager(AbstractProtocol networkProtocol, IKuzzle kuzzle) {
+      this.NetworkProtocol = networkProtocol;
+      networkProtocol.StateChanged += this.StateChangeListener;
+      this.kuzzle = kuzzle;
+      InitComponents();
+      this.kuzzle.GetEventHandler().UserLoggedIn += this.OnUserLoggedIn;
+    }
+
+    internal virtual void InitComponents() {
+      QueryReplayer = new QueryReplayer(this, kuzzle);
+      SubscriptionRecoverer = new SubscriptionRecoverer(this, kuzzle);
+      TokenVerifier = new TokenVerifier(this, kuzzle);
+    }
+
+    /// <summary>
+    /// Gets the network protocol.
+    /// </summary>
+    /// <returns>The network protocol.</returns>
+    internal override AbstractProtocol NetworkProtocol { get; set; }
+
+    /// <summary>
+    /// Gets the query replayer.
+    /// </summary>
+    /// <returns>The query replayer.</returns>
+    internal override IQueryReplayer QueryReplayer { get; set; }
+
+    /// <summary>
+    /// Gets the token verifier.
+    /// </summary>
+    /// <returns>The token verifier.</returns>
+    internal override ITokenVerifier TokenVerifier { get; set; }
+
+    /// <summary>
+    /// This will check the token validity,
+    /// and chose what to do before replaying the Queue
+    /// </summary>
+    internal override async Task Recover() {
+      if (!AutoRecover) return;
+
+      if (await TokenVerifier.IsTokenValid()) {
+        QueryReplayer.ReplayQueries();
+        QueryReplayer.Lock = false;
+        SubscriptionRecoverer.RenewSubscriptions();
+        return;
+      }
+
+      if (QueryReplayer.Lock) {
+        QueryReplayer.ReplayQueries((obj) =>
+        obj["controller"] != null
+        && obj["action"] != null
+        && obj["controller"].ToString() == "auth"
+        && obj["action"].ToString() == "login", false);
+      }
+
+    }
+
+    /// <summary>
+    /// This is used to verify if the user that has logged in
+    /// is the same that before, if not this will Reject every query in the Queue
+    /// and clear all subscriptions, otherwise this will replay the Queue if she is waiting.
+    /// </summary>
+    internal override void OnUserLoggedIn(object sender, UserLoggedInEvent e) {
+
+      if (previousKUID != e.Kuid) {
+
+        if (AutoRecover
+           && QueryReplayer.WaitLoginToReplay) {
+          QueryReplayer.RejectAllQueries(new UnauthorizeException("Unauthorized", 0));
```

Comment:
A more explicit message can be useful here, as it can be pretty difficult for users to understand why their request have been rejected.

```suggestion
          QueryReplayer.RejectAllQueries(new UnauthorizeException("Request submitted by another user", 0));
```